File: backend/app/schemas/attendance.py
"""Attendance Pydantic Schemas

考勤相关 Pydantic 模型
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional, List

from pydantic import BaseModel, Field, ConfigDict


class AttendanceBase(BaseModel):
    """考勤基础模型"""
    schedule_id: int = Field(..., description="排课ID")
    student_id: int = Field(..., description="学员ID")
    contract_id: Optional[int] = Field(None, description="合同ID(扣课时用)")
    status: int = Field(..., ge=1, le=4, description="状态: 1:出勤 2:请假 3:缺勤 4:迟到")
    check_time: Optional[datetime] = Field(None, description="签到时间")
    check_method: int = Field(default=1, ge=1, le=3, description="签到方式: 1:手动 2:人脸 3:刷卡")
    hours_consumed: Decimal = Field(default=Decimal("1.0"), gt=0, description="消耗课时数")
    notes: Optional[str] = Field(None, max_length=500, description="备注")


class AttendanceCreate(AttendanceBase):
    """考勤创建模型"""
    pass


class AttendanceUpdate(BaseModel):
    """考勤更新模型"""
    schedule_id: Optional[int] = Field(None, description="排课ID")
    student_id: Optional[int] = Field(None, description="学员ID")
    contract_id: Optional[int] = Field(None, description="合同ID")
    status: Optional[int] = Field(None, ge=1, le=4, description="状态: 1:出勤 2:请假 3:缺勤 4:迟到")
    check_time: Optional[datetime] = Field(None, description="签到时间")
    check_method: Optional[int] = Field(None, ge=1, le=3, description="签到方式: 1:手动 2:人脸 3:刷卡")
    hours_consumed: Optional[Decimal] = Field(None, gt=0, description="消耗课时数")
    notes: Optional[str] = Field(None, max_length=500, description="备注")


class AttendanceResponse(AttendanceBase):
    """考勤响应模型"""
    id: int
    created_by: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AttendanceBatchCreate(BaseModel):
    """批量创建考勤模型"""
    attendances: List[AttendanceCreate] = Field(..., min_length=1, description="考勤列表")
    auto_deduct_hours: bool = Field(default=True, description="是否自动扣减课时")


class AttendanceStatistics(BaseModel):
    """考勤统计模型"""
    total_count: int = Field(..., description="总考勤次数")
    present_count: int = Field(..., description="出勤次数")
    leave_count: int = Field(..., description="请假次数")
    absent_count: int = Field(..., description="缺勤次数")
    late_count: int = Field(..., description="迟到次数")
    present_rate: float = Field(..., description="出勤率(%)")
    total_hours_consumed: Decimal = Field(..., description="总消耗课时数")

File: miniapp/src/api/attendance.ts
/**
 * 考勤相关API
 */
import { request } from './request'

/**
 * 考勤状态枚举
 */
export enum AttendanceStatus {
  PRESENT = 1,   // 出勤
  LEAVE = 2,     // 请假
  ABSENT = 3,    // 缺勤
  LATE = 4       // 迟到
}

/**
 * 考勤状态文本映射
 */
export const AttendanceStatusText: Record<AttendanceStatus, string> = {
  [AttendanceStatus.PRESENT]: '出勤',
  [AttendanceStatus.LEAVE]: '请假',
  [AttendanceStatus.ABSENT]: '缺勤',
  [AttendanceStatus.LATE]: '迟到'
}

/**
 * 考勤状态颜色映射
 */
export const AttendanceStatusColor: Record<AttendanceStatus, string> = {
  [AttendanceStatus.PRESENT]: '#52c41a',
  [AttendanceStatus.LEAVE]: '#1890ff',
  [AttendanceStatus.ABSENT]: '#ff4d4f',
  [AttendanceStatus.LATE]: '#faad14'
}

/**
 * 考勤记录
 */
export interface Attendance {
  id: number
  schedule_id: number
  schedule?: {
    id: number
    course_name: string
    start_time: string
    end_time: string
    classroom_name: string
  }
  student_id: number
  contract_id: number | null
  status: AttendanceStatus
  check_time: string | null
  check_method: number
  hours_consumed: number
  notes: string | null
  created_by: number | null
  created_at: string
}

/**
 * 考勤创建参数
 */
export interface AttendanceCreateParams {
  schedule_id: number
  student_id: number
  contract_id?: number
  status: AttendanceStatus
  check_time?: string
  check_method?: number
  hours_consumed?: number
  notes?: string
}

/**
 * 考勤更新参数
 */
export interface AttendanceUpdateParams {
  schedule_id?: number
  student_id?: number
  status?: AttendanceStatus
  check_time?: string
  check_method?: number
  hours_consumed?: number
  notes?: string
}

/**
 * 考勤列表查询参数
 */
export interface AttendanceListParams {
  skip?: number
  limit?: number
  schedule_id?: number
  student_id?: number
  contract_id?: number
  status?: AttendanceStatus
  start_date?: string
  end_date?: string
}

/**
 * 考勤统计
 */
export interface AttendanceStatistics {
  total_count: number
  present_count: number
  leave_count: number
  absent_count: number
  late_count: number
  present_rate: number
  total_hours_consumed: number
}

/**
 * 考勤API类
 */
export class AttendanceAPI {
  /**
   * 获取考勤列表
   */
  static async list(params?: AttendanceListParams): Promise<Attendance[]> {
    return request.get<Attendance[]>('/attendances', params)
  }

  /**
   * 获取考勤详情
   */
  static async getById(attendanceId: number): Promise<Attendance> {
    return request.get<Attendance>(`/attendances/${attendanceId}`)
  }

  /**
   * 创建考勤记录
   */
  static async create(data: AttendanceCreateParams): Promise<Attendance> {
    return request.post<Attendance>('/attendances', data)
  }

  /**
   * 更新考勤记录
   */
  static async update(attendanceId: number, data: AttendanceUpdateParams): Promise<Attendance> {
    return request.put<Attendance>(`/attendances/${attendanceId}`, data)
  }

  /**
   * 删除考勤记录
   */
  static async delete(attendanceId: number): Promise<void> {
    return request.delete<void>(`/attendances/${attendanceId}`)
  }

  /**
   * 获取学员考勤统计
   */
  static async getStudentStatistics(studentId: number): Promise<AttendanceStatistics> {
    return request.get<AttendanceStatistics>(`/attendances/statistics/student/${studentId}`)
  }

  /**
   * 获取排课考勤统计
   */
  static async getScheduleStatistics(scheduleId: number): Promise<AttendanceStatistics> {
    return request.get<AttendanceStatistics>(`/attendances/statistics/schedule/${scheduleId}`)
  }

  /**
   * 获取课程考勤统计
   */
  static async getCourseStatistics(
    courseId: number,
    startDate?: string,
    endDate?: string
  ): Promise<AttendanceStatistics> {
    return request.get<AttendanceStatistics>(`/attendances/statistics/course/${courseId}`, {
      start_date: startDate,
      end_date: endDate
    })
  }

  /**
   * 获取我的考勤记录
   */
  static async getMyAttendances(params?: Partial<AttendanceListParams>): Promise<Attendance[]> {
    return this.list({
      ...params
      // student_id会自动从当前用户信息中获取
    })
  }

  /**
   * 统计考勤数量
   */
  static async count(params?: Partial<AttendanceListParams>): Promise<{ count: number }> {
    return request.get('/attendances/count', params)
  }
}

export default AttendanceAPI
